### PR TITLE
feat: lazy iterator, less allocation

### PR DESCRIPTION
**Pull request by KSXGitHub**

### CODE REVIEW

**Review by anonrig — APPROVED** [review]
How does this effect the performance? Can you run the benchmark folder and share before after execution times of adding fastify?

File: crates/cli/src/commands/install.rs [hunk 1]
```diff
@@ -36,12 +36,11 @@ impl PackageManager {
         let config = &self.config;
         let path = &self.config.modules_dir;
         let http_client = &self.http_client;
-        let mut queue: VecDeque<Vec<PackageVersion>> = VecDeque::new();
+        let mut queue: VecDeque<Vec<PackageVersion>> = VecDeque::new(); // QUESTION: is this queue necessary since it only has one element?
```
Comment (anonrig) [inline crates/cli/src/commands/install.rs:39]: Initially there is one, but this package might have multiple dependencies as well. So queue is important to iterate and install all of them.

Comment (KSXGitHub) [inline crates/cli/src/commands/install.rs:39]:
```suggestion
        let mut queue: VecDeque<Vec<PackageVersion>> = VecDeque::new();
```

**Comment by KSXGitHub** [comment]
> How does this effect the performance? Can you run the benchmark folder and share before after execution times of adding fastify?

**main (2980669):**

benchmark screenshot of main

**this PR (54939f4):**

benchmark screenshot of this PR

Speaking of benchmark, I have a repo called parallel-disk-usage which compare benchmark of the PR against a fixed version: https://github.com/KSXGitHub/parallel-disk-usage/blob/0.9.0/.github/workflows/benchmark.yaml

**Comment by anonrig** [comment]
@KSXGitHub I mean can you run [this benchmark](https://github.com/pnpm/pacquet/blob/main/benchmark/add.sh)?

**Comment by KSXGitHub** [comment]
The warmup=5 is too much for me. I think I will benchmark it tomorrow. Or you can benchmark it yourself. Though I don't think there would be any visible difference, as the cost of allocations is negligible compared to network latency and IO.

**Comment by KSXGitHub** [comment]
BTW, I have a repo called parallel-disk-usage that has a CI that compare benchmark of PRs to a fixed version: https://github.com/KSXGitHub/parallel-disk-usage/blob/0.9.0/.github/workflows/benchmark.yaml. The deploy CI also benchmark it against similar tool: https://github.com/KSXGitHub/parallel-disk-usage/blob/0.9.0/.github/workflows/deploy.yaml.

**Pull request merged**